File: python/flink_agents/runtime/memory/event_attachment_utils.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
#################################################################################
from __future__ import annotations

import hashlib
from typing import TYPE_CHECKING, Any

from flink_agents.api.events.event import OutputEvent
from flink_agents.api.memory_object import MemoryType, validate_memory_value
from flink_agents.api.memory_reference import MemoryRef

if TYPE_CHECKING:
    from uuid import UUID

    from flink_agents.api.events.event import Event
    from flink_agents.api.runner_context import RunnerContext

_ATTACHMENT_ROOT = "__event_attachments__"


class EventAttachmentError(RuntimeError):
    """Raised when an Event attachment cannot be stored or loaded."""


def _hash_attachment_key(key: str) -> str:
    return hashlib.sha256(key.encode("UTF-8")).hexdigest()


def build_attachment_path(event_id: UUID, key: str) -> str:
    """Build the canonical SensoryMemory path for one attachment."""
    return f"{_ATTACHMENT_ROOT}.{event_id}.{_hash_attachment_key(key)}"


def _attachment_context(event: Event, key: str, path: str | None = None) -> str:
    suffix = f", path={path}" if path is not None else ""
    return f"event_id={event.id}, event_type={event.type}, key={key}{suffix}"


def store_event_attachments(event: Event, ctx: RunnerContext) -> None:
    """Store concrete attachment values in SensoryMemory and replace them with refs."""
    if not event.attachments:
        return

    if event.type == OutputEvent.EVENT_TYPE:
        keys = ", ".join(sorted(event.attachments))
        msg = f"Output events cannot carry attachments: {_attachment_context(event, keys)}"
        raise EventAttachmentError(msg)

    pending: list[tuple[str, str, Any]] = []
    for key, value in event.attachments.items():
        if isinstance(value, MemoryRef):
            if value.memory_type != MemoryType.SENSORY:
                msg = (
                    "Event attachments must use sensory memory references: "
                    f"{_attachment_context(event, key, value.path)}"
                )
                raise EventAttachmentError(msg)
            continue
        path = build_attachment_path(event.id, key)
        try:
            validate_memory_value(path, value)
        except Exception as exc:
            msg = f"Invalid event attachment value: {_attachment_context(event, key, path)}"
            raise EventAttachmentError(msg) from exc
        pending.append((key, path, value))

    for key, path, value in pending:
        event.attachments[key] = ctx.sensory_memory.set(path, value)


def load_event_attachments(event: Event, ctx: RunnerContext) -> None:
    """Load sensory refs in place immediately before a Python Action runs."""
    for key, value in list(event.attachments.items()):
        if not isinstance(value, MemoryRef):
            continue

        try:
            sensory_memory = ctx.sensory_memory
            attachment = sensory_memory.get(value)
            attachment_missing = attachment is None and not sensory_memory.is_exist(
                value.path
            )
        except Exception as exc:
            msg = f"Failed to load event attachment: {_attachment_context(event, key, value.path)}"
            raise EventAttachmentError(msg) from exc

        if attachment_missing:
            msg = (
                "Event attachment does not exist in sensory memory: "
                f"{_attachment_context(event, key, value.path)}"
            )
            raise EventAttachmentError(msg)
        event.attachments[key] = attachment

File: runtime/src/main/java/org/apache/flink/agents/runtime/memory/EventAttachmentUtils.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.flink.agents.runtime.memory;

import org.apache.flink.agents.api.Event;
import org.apache.flink.agents.api.OutputEvent;
import org.apache.flink.agents.api.context.MemoryObject;
import org.apache.flink.agents.api.context.MemoryRef;
import org.apache.flink.agents.api.context.RunnerContext;
import org.apache.flink.api.common.typeutils.TypeSerializer;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Map;
import java.util.Objects;
import java.util.UUID;
import java.util.stream.Collectors;

/** Stores event attachments in sensory memory while events cross action boundaries. */
public final class EventAttachmentUtils {

    private static final String ATTACHMENT_ROOT = "__event_attachments__";

    private EventAttachmentUtils() {}

    /** Stores concrete attachment values and replaces them with sensory-memory references. */
    public static void storeEventAttachments(Event event, RunnerContext context) throws Exception {
        if (event.getAttachments().isEmpty()) {
            return;
        }

        if (OutputEvent.EVENT_TYPE.equals(event.getType())) {
            String keys =
                    event.getAttachments().keySet().stream()
                            .sorted()
                            .collect(Collectors.joining(", "));
            throw new IllegalArgumentException(
                    "Output events cannot carry attachments: event_id="
                            + event.getId()
                            + ", event_type="
                            + event.getType()
                            + ", key="
                            + keys);
        }

        for (Map.Entry<String, Object> entry : event.getAttachments().entrySet()) {
            String key = entry.getKey();
            Object value = entry.getValue();
            if (value instanceof MemoryRef) {
                MemoryRef reference = (MemoryRef) value;
                if (!MemoryObject.MemoryType.SENSORY.equals(reference.getType())) {
                    throw new IllegalArgumentException(
                            "Event attachments must use sensory memory references: event_id="
                                    + event.getId()
                                    + ", event_type="
                                    + event.getType()
                                    + ", key="
                                    + key
                                    + ", memory_type="
                                    + reference.getType());
                }
                continue;
            }

            MemoryRef reference =
                    context.getSensoryMemory().set(buildAttachmentPath(event.getId(), key), value);

            event.getAttachments().put(key, reference);
        }
    }

    /** Returns an Event with sensory-memory references resolved on a copy when necessary. */
    public static Event loadEventAttachments(
            Event event, RunnerContext context, TypeSerializer<Event> eventSerializer)
            throws Exception {
        boolean requiresResolution =
                event.getAttachments().values().stream().anyMatch(MemoryRef.class::isInstance);
        if (!requiresResolution) {
            return event;
        }

        Event actionEvent =
                Objects.requireNonNull(
                                eventSerializer,
                                "Event serializer is required to resolve attachment references.")
                        .copy(event);
        for (Map.Entry<String, Object> entry : actionEvent.getAttachments().entrySet()) {
            Object value = entry.getValue();
            if (!(value instanceof MemoryRef)) {
                continue;
            }
            MemoryRef reference = (MemoryRef) value;

            MemoryObject attachment = context.getSensoryMemory().get(reference);
            if (attachment == null) {
                throw new IllegalStateException(
                        "Event attachment does not exist in sensory memory: "
                                + reference.getPath());
            }
            actionEvent.getAttachments().put(entry.getKey(), attachment.getValue());
        }
        return actionEvent;
    }

    /** Builds the sensory-memory path for one event attachment. */
    public static String buildAttachmentPath(UUID eventId, String key) {
        if (eventId == null) {
            throw new IllegalArgumentException("Event attachment requires a non-null event id.");
        }
        if (key == null) {
            throw new IllegalArgumentException("Event attachment key must not be null.");
        }
        return ATTACHMENT_ROOT + "." + eventId + "." + hashAttachmentKey(key);
    }

    private static String hashAttachmentKey(String key) {
        try {
            byte[] digest =
                    MessageDigest.getInstance("SHA-256")
                            .digest(key.getBytes(StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder(digest.length * 2);
            for (byte value : digest) {
                sb.append(String.format("%02x", value));
            }
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("SHA-256 is not available.", e);
        }
    }
}
